# Portfolio and Risk Management
## Homework 1: HMC and Inflation-Indexed Bonds
### Group A


In [ ]:
import pandas as pd 
path = "data/multi_asset_etf_data.xlsx"
excess_returns = pd.read_excel(path, sheet_name="excess returns", index_col=0)
excess_returns = excess_returns.drop(columns = "QAI")

# 2 Mean-Variance Optimization
## 2.1 Summary Statistics

In [ ]:
import numpy as np
annualized_mean_returns = excess_returns.mean() * 12
annualized_return_volatility = excess_returns.std() * np.sqrt(12)
sharpe_ratio = annualized_mean_returns / annualized_return_volatility
summary_stats = pd.DataFrame({"annualized_mean": annualized_mean_returns, "annualized volatility": annualized_return_volatility, "sharpe ratio": sharpe_ratio})
print(summary_stats)

     annualized_mean  annualized volatility  sharpe ratio
BWX        -0.017163               0.081891     -0.209582
DBC         0.013000               0.172276      0.075461
EEM         0.044426               0.179085      0.248074
EFA         0.063754               0.148590      0.429057
HYG         0.034500               0.073326      0.470501
IEF         0.008596               0.062120      0.138377
IYR         0.070095               0.165531      0.423456
PSP         0.076428               0.211323      0.361666
SPY         0.126955               0.141398      0.897856
TIP         0.012776               0.049748      0.256817


### Which assets have the best and worse Sharpe Ratios?

The best Sharpe Ratios is the highest one since that represents more return for a certain level of risk, therefore the asset with the best Sharpe Ratio is SPY with a ratio of 0.898. The worst ratio is the lowest ratio which is BWX with the Sharpe Ratio of -0.20958. As you can see in the table above, BWX has a negative annualized mean which is what causes the ratio to be negative. 

## 2.2 Descriptive Analysis

In [ ]:
correlation_matrix = excess_returns.corr()
print(correlation_matrix)

          BWX       DBC       EEM       EFA       HYG       IEF       IYR  \
BWX  1.000000  0.160774  0.611691  0.612692  0.600427  0.580844  0.555010   
DBC  0.160774  1.000000  0.438794  0.440497  0.418825 -0.308778  0.252598   
EEM  0.611691  0.438794  1.000000  0.811959  0.669611  0.051462  0.574328   
EFA  0.612692  0.440497  0.811959  1.000000  0.776442  0.062915  0.700449   
HYG  0.600427  0.418825  0.669611  0.776442  1.000000  0.192726  0.731282   
IEF  0.580844 -0.308778  0.051462  0.062915  0.192726  1.000000  0.321839   
IYR  0.555010  0.252598  0.574328  0.700449  0.731282  0.321839  1.000000   
PSP  0.523903  0.420190  0.697394  0.861604  0.798316  0.012493  0.726034   
SPY  0.443512  0.383582  0.693457  0.832782  0.783147  0.011276  0.743470   
TIP  0.675024  0.084279  0.383897  0.402497  0.540414  0.756082  0.599880   

          PSP       SPY       TIP  
BWX  0.523903  0.443512  0.675024  
DBC  0.420190  0.383582  0.084279  
EEM  0.697394  0.693457  0.383897  
EFA  0.8

In [ ]:
asset_classes = correlation_matrix.columns
result = []


for i in range(len(asset_classes)):
    for j in range(i + 1, len(asset_classes)):        
        
        asset1 = asset_classes[i]
        asset2 = asset_classes[j]
        selected_corr = correlation_matrix.loc[asset1, asset2]

        result.append((asset1, asset2, selected_corr))

asset_pairs = pd.DataFrame(result, columns = ["asset1", "asset2", "correlation coefficient"]).sort_values("correlation coefficient", key=abs)
# print(asset_pairs)
highest = asset_pairs.head(1)
lowest = asset_pairs.tail(1)
print(highest), print(lowest)

   asset1 asset2  correlation coefficient
37    IEF    SPY                 0.011276
   asset1 asset2  correlation coefficient
42    PSP    SPY                 0.872151


(None, None)

Highest correlation is between IEF and SPY with a correlation coefficient of 0.011276
Lowest Correlation is between PSP and SPY with a correlation coefficient of 0.8722


In [ ]:
print(summary_stats.loc[["TIP", "IEF", "BWX"]])
print(correlation_matrix.loc["TIP", ["IEF", "BWX"]])

     annualized_mean  annualized volatility  sharpe ratio
TIP         0.012776               0.049748      0.256817
IEF         0.008596               0.062120      0.138377
BWX        -0.017163               0.081891     -0.209582
IEF    0.756082
BWX    0.675024
Name: TIP, dtype: float64


In our sample, TIPS outperforms Domestic Bonds and Foreign Bonds

## 2.3 The MV Frontier

## 2.4 TIPS

# 3 Allocations